### DISCUSSION TITLE
Default run infeasibilities in v4.3.5 q35_land_other(LAM_80)

### DISCUSSION THREAD

**Discussion by jansteinhauser** [magpiemodel/magpie#333] [Q&A] [answered]
Hello again!

I've just upgraded to MAgPIE v4.3.5, but sadly I can't seem to get it to run properly. Even with a clean install without any of my project-specific changes the default run stops due to an infeasibility:

```
WARNING: Infeasible solutions found (2020, 2025, 2030, 2035, 2040, 2045, 2050, 2055, 2060, 2070, 2080, 2090, 2100)!
```

The lst gives me the following:
```
An equation in the pre-triangular part of the model cannot be solved because the critical variable is at a bound.
Residual = 3.26475145E-02
Tolerance (Tol_Feas_Tria) = 1.00000000E-08
Equation q35_min_other(LAM_80) turned into simple lower bound on Variable vm_land(LAM_80,other). Bound = 0.120196
Equation q35_other_conversion(LAM_80) turned into simple lower bound on Variable v35_other(LAM_80,ac0). Bound = 0.110616
Variable vm_bv(LAM_80,other,nonforested) is forced to its lower bound = 0
Variable v35_other(LAM_80,acx) is forced to its upper bound = 0.0117809
Variable v35_other(LAM_80,ac25) is forced to its upper bound = 0.000181084
Variable v35_other(LAM_80,ac15) is forced to its upper bound = 0.030151
Variable v35_other(LAM_80,ac5) is forced to its upper bound = 0.000114557
Variable v35_other(LAM_80,ac0) is forced to its current value = 0.110616
Equation q35_bv_other(LAM_80,nonforested) acting as binding =L= has forced the above variables.
Equation q35_land_other(LAM_80) cannot be solved with respect to Variable vm_land(LAM_80,other) due to an active bound. Value = 0.120196  
Residual after last transformation = 0.032648 
```

R and R packages are on their respective newest versions, GAMS is v35.2.0. 

R also gives the following warnings, but those don't seem to be actual issues as none of the files are used:
```
1: In copy_input(x = file2destination, sourcepath = "input", suffix = low_res,  :  File avl_cropland.cs3 seems to be missing!
2: In copy_input(x = file2destination, sourcepath = "input", suffix = low_res,  :  File avl_cropland_0.5.mz seems to be missing!
3: In copy_input(x = file2destination, sourcepath = "input", suffix = low_res,  :  File f38_historical_share.csv seems to be missing!
4: In copy_input(x = file2destination, sourcepath = "input", suffix = low_res,  :  File f38_regression_cap_share.csv seems to be missing!
5: In copy_input(x = file2destination, sourcepath = "input", suffix = low_res,  :  File f38_fac_req_fao.csv seems to be missing!
```

Is this a known issue of the current release or something unique to my default runs? Any suggestions what to fix where? Or should I simply revert to v4.3.4 - obviously not my preferred solution?

Thanks and all the best!
Jan

**Comment by abhimishr** [comment, marked as answer]
Possible solution here : https://github.com/magpiemodel/magpie/discussions/333#discussioncomment-1558237

--------------------------------------------------
To replicate the issue, can you confirm if the following fails :

1. Fresh clone of 4.3.5 release
2. Using the default start script

We always make sure that the default runs work as expected before every release. Additionally, if you paste the config.log from your failing runs we can compare it to a known 4.3.5 default run and see what's going on/what changed. 

I have a feeling that somehow the land protection policies are playing a role in your infeasibilities but this should not happen for a default run with freshly downloaded data.

**Reply by jansteinhauser** [reply]
Thanks for the quick response.
I did as you suggested and cloned once more into a new folder, ran default again, and had the same issue, sadly.

[config.log](https://github.com/magpiemodel/magpie/files/7376294/config.log)

I also went back to my old MAgPIE clone and ran a default run on that again just to be sure that it's not some general issue. It worked on that version.

Thank you for looking into this!

**Reply by abhimishr** [reply]
Thanks! I could not find any discernible differences in the config compared to a known run except for some additional info towards the end of config but I think it is not related to the infeasibilities. I will setup a run here and see if the infeasibilty shows up again. And just to be sure again, you are running the `master` branch?

Edit: Just noticed one change where you seem to have 
```
$info
$info$version
[1] "4.3.4dev"
```
which for the release version should be 
```
$info
$info$version
[1] "4.3.5"
```
Maybe a bad merge somewhere?

**Reply by abhimishr** [reply]
Hi Jan,

My run just finished with all time steps being feasible. I am running GAMS 36.2.0 - but this should not matter as we were running GAMS 35.x for a long time - also when version 4.3.5 released. 

```
--- LOOPS t = y2100
---       FOR/WHILE = 1
---   87,400 rows  87,649 columns  265,918 non-zeroes
---   195,313 nl-code  46,050 nl-non-zeroes
--- full.gms(1040141) 1719 Mb
--- Executing CONOPT4 (Solvelink=2): elapsed 0:28:27.864

CONOPT 4         36.2.0 r433180e Released Sep 03, 2021 LEG x86 64bit/Linux

    The user model has 87400 constraints and 87649 variables
    with 265918 Jacobian elements, 46050 of which are nonlinear.
    The Hessian of the Lagrangian has 249 elements on the diagonal,
    7719 elements below the diagonal, and 9213 nonlinear variables.

    Iter Phase   Ninf   Infeasibility   RGmax      NSB   Step  InItr MX OK
       0   0          2.2590765265E+05 (Input point)

    The pre-triangular part of the model has 3880 constraints and variables.
    The post-triangular part of the model has 1 constraints and variables.
    There are 72182 definitional constraints and defined variables.

    Preprocessed model has 11337 constraints and 11586 variables
    with 145290 Jacobian elements, 129339 of which are nonlinear.

    Iter Phase   Ninf   Infeasibility   RGmax      NSB   Step  InItr MX OK
                      2.2590765265E+05 (Full preprocessed model)
                      9.6594697213E+00 (After scaling)
                      9.3488738692E+00 (After adjusting individual variables)
       1   0       0  9.2674571407E+00                 1.0E+00   221 T  T
       2   1     195  5.1885509902E+00 8.6E+02      25 1.0E+00    25 T  T
       3   1     145  3.0960789377E+00 1.2E+02      50 1.0E+00    50 T  T
       4   1      45  6.4104559427E-01 3.0E+01     100 1.0E+00   100 T  T
       5   1       7  9.2036165155E-02 1.5E+01      38 1.0E+00    38 T  T
       6   1       1  7.6802716099E-03 1.5E+01       6 1.0E+00     6 T  T

 ** Feasible solution. Value of objective =   -1686.95587747

    Iter Phase   Ninf     Objective     RGmax      NSB   Step  InItr MX OK
       7   3         -0.0000000000E+00 1.0E+06      10 1.0E+00     1 T  T
       8   4         -0.0000000000E+00 0.0E+00       0

 ** Optimal solution. There are no superbasic variables.

--- Reading solution for model m15_food_demand
--- GDX Point /p/projects/magpie/users/mishra/projects/magpiemodel/magpiemodel_master/output/default_2021-10-20_06.28.27/m15_food_demand_p.gdx
--- Executing after solve: elapsed 0:28:30.962
--- full.gms(1033619) 1738 Mb
--- GDX File (execute_unload) /p/projects/magpie/users/mishra/projects/magpiemodel/magpiemodel_master/output/default_2021-10-20_06.28.27/fulldata.gdx
--- Putfile dummy /p/projects/magpie/users/mishra/projects/magpiemodel/magpiemodel_master/output/default_2021-10-20_06.28.27/dummy.put
*** Status: Normal completion
--- Job full.gms Stop 10/20/21 06:59:10 elapsed 0:28:35.645
```

**Reply by abhimishr** [reply]
Tagging @tscheypidi - How do we save values in cfg$info? Its declared as a list in the config but somewhere down the line it is populated with version info. Jan's version seems to be `"4.3.4dev"` but it should be `"4.3.5"` for latest release. Do you know if a unclean merge would cause this? (Jan says that the issue remains even on a fresh clone so seems bit unlikely)

Meanwhile, @jansteinhauser, can you repeat the following steps?

1.Create a new directory, make a folder called `master` then run the following (while being in the new directory but not inside the newly created `master` folder)
`git clone -b master git@github.com:magpiemodel/magpie.git master`
2. Navigate to the `master` folder
3. Open a command prompt/terminal in this location
4. Run `Rscript start.R`
5. Choose default (usually option `1`)
6. Choose direct execution if running on a personal computer (on a cluster this might look different)

**Reply by jansteinhauser** [reply]
Thank you for the quick and thorough help!

> 1.Create a new directory, make a folder called `master` then run the following (while being in the new directory but not inside the newly created `master` folder)
> `git clone -b master git@github.com:magpiemodel/magpie.git master`
> 2. Navigate to the `master` folder
> 3. Open a command prompt/terminal in this location
> 4. Run `Rscript start.R`
> 5. Choose default (usually option `1`)
> 6. Choose direct execution if running on a personal computer (on a cluster this might look different)

Did so - but to no avail. Again infeas starting in 2020.
I'm not sure it might have made a difference, but I wasn't able to use ` git clone -b master git@github.com:magpiemodel/magpie.git master` directly, because:

`
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.
`

Instead, I used `git clone -b master https://github.com/magpiemodel/magpie.git master`. So, as far as I can tell, essentially the same command. 

Most recent config log, now showing v4.3.5: 
[config.log](https://github.com/magpiemodel/magpie/files/7380045/config.log)

I guess I'll experiment with this on a second PC as well to see if that solves it. And, just for the sake of it, update my GAMS to the latest version as well.

**Reply by abhimishr** [reply]
One more place where we can make a check - can you share a full.gms file from failing runs - then we get a better idea of what has changed in totality (as the full.gms file would contain all the settings,model specifications and input data used)

You can compress/zip the file as the `full.gms` in itself is quite big.

**Reply by jansteinhauser** [reply]
As expected, updating GAMS to v36 didn't change anything.

Also, here's the most recent full.gms (it won't upload 7z, so it's compressed twice - in case you're wondering):
[full.zip](https://github.com/magpiemodel/magpie/files/7380720/full.zip)

Thanks again for taking the time to look at this.

**Reply by abhimishr** [reply]
Thanks for the zipped gms file. The file I have and the one you have are identical (Except for a gms package warning in your case L209:210, but i think this should not cause infeasibilities.

```
 Warning message:
 package 'gms' was built under R version 4.1.1
```

Strange that the model is somehow still infeasible. Can you also post your full.log file? I would like to see the initial point from where the issue starts. Meanwhile, maybe try updating the `gms` package and make a fresh run if possible. Once you update `gms` package, make sure that it loads correctly via `library(gms)` and does not throw any warning.

**Reply by jansteinhauser** [reply]
Good morning!
Thanks again for your suggestions.

It's strange. My gms is v0.10.4  and I run R v4.1.1; manually loading gms via `library(gms)` throws no warning. Yet, full.gms still has the warning message about 4.1.1. Additionally strange: Before upgrading to 4.1.1, I got similar warnings for several packages, including gms, that were built under 4.1.1 - but already when starting the script. For gms no warning pops up while the script runs and, if there still was an issue with my R version, I´d expect more warnings in the log as well. Any idea why this behavior may be limited to gms and the log?

Here is a new zip with config.log, full.log and full gms from a new run:
[default_2021-10-21_08.40.28.zip](https://github.com/magpiemodel/magpie/files/7387187/default_2021-10-21_08.40.28.zip)

Also, I've set up MAgPIE v4.3.5 on a second PC with a completely clean install of GAMS 36 and R v4.1.1. No gms warning, but still infeasibilities starting in 2020.

I can't imagine it but since I want to remove all potential error sources, could this be an issue:
> I'm not sure it might have made a difference, but I wasn't able to use ` git clone -b master git@github.com:magpiemodel/magpie.git master` directly, because:
> `git@github.com: Permission denied (publickey). fatal: Could not read from remote repository.`
> Instead, I used git clone -b master https://github.com/magpiemodel/magpie.git master. So, as far as I can tell, essentially the same command.

Thanks again for your help

**Reply by abhimishr** [reply]
> I can't imagine it but since I want to remove all potential error sources, could this be an issue:

No this is basically not letting you clone this repo via the ssh key (possibly to avoid non-PIK accounts pushing code to central repo).

From the log file I can see that the issue comes from L1039737 in full.gms (its little bit offset in line numbers here and there). The portion where this error happens is turning off some CONOPT4 pre-processing when modelstat is 13. Maybe try commenting out this chunk from solve.gms of the default optimization realization
```
* if solve stopped with an error, try it again without pre-processing
    if((magpie.modelstat = 13),
      display "WARNING: Modelstat 13 | retry without Conopt4 pre-processing";
	  magpie.optfile = 2
      solve magpie USING nlp MINIMIZING vm_cost_glo;
      magpie.optfile   = s80_optfile ;
    );
```
We have another if comment below to catch modelstat 13 situations. This is just a guess as modelstat 13 already in 2020 is a problem in its own but lets try.

**Reply by abhimishr** [reply]
Also, can you maybe replace the contents of default start script with this?
```
# Load start_run(cfg) function which is needed to start MAgPIE runs
source("scripts/start_functions.R")
source("config/default.cfg")

#start MAgPIE run
cfg$results_folder = "output/:title:"
cfg$title = "MAgPIE4p5_Test1"
cfg <- setScenario(cfg,c("BASE"))
start_run(cfg)
```
Just to turn off all additional land protection and see if thats the culprit

**Reply by jansteinhauser** [reply]
Commenting out the conditional for modelstat 13 didn't do anything, but good news: The second one did the trick! I did actually get through a full run that way. 
Still strange that there's an issue with the "true" default that doesn't occur for others.

**Reply by abhimishr** [reply]
> Commenting out the conditional for modelstat 13 didn't do anything, but good news: The second one did the trick! I did actually get through a full run that way. Still strange that there's an issue with the "true" default that doesn't occur for others.

[Looks like indeed something going strange with land protection](https://github.com/magpiemodel/magpie/discussions/333#discussioncomment-1504350) as BASE run turns off the NPI setting. At this stage I am bit out of ideas but we have a possible culprit. Tagging @tscheypidi @flohump and @mishkos to have better suggestions.

**Reply by jansteinhauser** [reply]
Thanks so much for your help so far! Enjoy your weekend :)

**Reply by devenazevedo** [reply]
Hi both, as discussed earlier today, I tried running a default run on a fresh clone of 4.3.5 on my Windows machine and also had the model stop due to an infeasibility in the year 2020. So indeed seems likely to be related to Windows OS! Attached is the full.lst in case useful. Let me know if I can help in any other way!
[full.zip](https://github.com/magpiemodel/magpie/files/7429080/full.zip)

**Reply by abhimishr** [reply]
Interesting! Thanks Deven. I will try to contact the conopt and gams developers and see if they have some insights.

**Reply by jansteinhauser** [reply]
Thanks to both of you! Weird if that truly is the issue but at least I can stop worrying that I failed at the basic installation or something along those lines. :) 
For now, I'll try the (seemingly) simple thing of switching to a Unix machine. Wish me luck!

**Reply by abhimishr** [reply]
Hi Jan and Deven,

My exchange with conopt team was bit of a mixed bag, but we understood that there are fundamental rounding differences in conopt/GAMS in Windows and Unix system; hence the solution paths are not identical. Anyhow, looks like windows model drops additional variables as compared to Unix models. Conopt developers suggested diving deeper into this and also recommended a different solver to diagnose the issue better, this I will do on our side.

Meanwhile, the good news is that I was able to update a few scaling numbers, and I was able to generate a feasible default magpie run on a Windows machine. I ask you both to do the following on a Windows machine:

1. in `10_land\landmatrix_dec18\scaling.gms` update `vm_landdiff.scale = 10e4;` as `vm_landdiff.scale = 10e3;`
2. in `32_forestry\dynamic_feb21\scaling.gms` add `v32_land_expansion.scale(j,type32,ac) = 10e-3;` and `vm_cdr_aff.scale(j,ac,aff_effect) = 10e-4;`
3. in `35_natveg\dynamic_feb21\scaling.gms` update `vm_landdiff_natveg.scale = 10e4;` as `vm_landdiff_natveg.scale = 10e3;`
4. in `71_disagg_lvst\foragebased_aug18\scaling.gms` add `v71_feed_balanceflow.scale(j,kli_rum,kforage) = 10e3;`

Can you set up a default magpie run with these changes and see if the model still fails in 2020?

**Reply by jansteinhauser** [reply]
Hello again and thanks again for all the effort. I did as you suggested and did indeed successfully run a default 4.3.5 run with your changes on my Windows machine. 
Did you by any chance compare the fulldata.gdx or selected outputs of your Windows run to one from a Unix run? Would be interesting to know if and how large the difference are.

Also, if you repost your last response as a top-level comment, we could mark it as (temporary?) solution for this issue.

Thanks again!

**Reply by jansteinhauser** [reply]
Oh, quick follow up question as this may also be a Windows issue: On neither of my Windows PCs and in none of the versions (4.0, 4.3.4, 4.3.5) did I ever get a proper validation report. My reports are roughly 60 pages or shorter; proper ones would be thousands of pages, wouldn't they? 

As before, I know that other (Unix-)PCs produce proper reports. So, maybe again an OS issue?

I can turn this into its own question, just figured it's a good fit here.
Thanks!

**Reply by devenazevedo** [reply]
Hi both, probably not all that useful now but just a quick note to say Abhijeet's changes to the scaling numbers also worked for me! I didn't run this with a the validation report turned on, so I'll rerun with it turned on and see if I also get this issue.

**Reply by devenazevedo** [reply]
Hey both, on the validation report, this again probably isn't super useful, but it looks like the validation report isn't compiling for me due to an error with miktex (at Vivid we usually use the shiny app or plot results ourselves to adjust formatting etc so I don't usually check the validation report). I assume this is just an issue with miktex on my machine so will try to solve this so I can be of more help, however, it may take a few weeks as we're having to migrate to new laptops this month so I'm reluctant to invest time into updating/installing new software on my old machine. Maybe it would be helpful to know that in the figures folder (from which I assume the validation report is compiled) inside the output folder, I have 2648 individual pdf's each with a separate plot - do you also have these plots and then it's just that the full report doesn't compile?

**Reply by tscheypidi** [reply]
about the Miktex issue: this is usally an indication that some latex packages are not installed yet. Miktex has a setting which allows to install these missing packages automatically. Turning that on should solve the problem

**Reply by flohump** [reply]
Just a quick update on that: In the upcoming MAgPIE 4.4.0 release we introduce a shorter version of the `validation.pdf` file as default, which is about half in size and pages by including only aggregate crop groups such as cereals, sugar and oil. The shorter version also reduces the time for creating the PDF by more than 50% (rather 15-20 minutes instead of 45 minutes).

**Reply by devenazevedo** [reply]
Thanks both! Embarrassingly the Miktex issue turned out to just be from Miktex needing to be updated, and now that that is done the validation report compiles fine for me, ending up with 2927 pages. So @jansteinhauser this doesn't seem to be a windows issue. And Florian, that sounds like a great idea! As implied above, I often don't run the validation report due to the time required (on my computer it takes closer to 1.5 hours)

**Reply by jansteinhauser** [reply]
It's the little things sometimes. :) 
Thanks for the update, good to know that it's working for you. I guess I'll have to keep digging for this then.
Thanks again!

**Reply by devenazevedo** [reply]
Haha indeed! No problem - let me know if I can help by testing anything on my windows machine in the future!

**Comment by jansteinhauser** [comment]
Since it came up in Wednesday's meeting, a quick update on the 4.4.0 RC: The default run does indeed solve without any infeasibilities on Windows. So, no problem there, it seems. (Still confused about my 51 page report, though. But that's another issue)

**Reply by flohump** [reply]
Good news and thanks for testing! Please note that the input data for rev4.65 on the RSE server is not yet final. There will be another update replacing the rev4.65 input files. Just to make you aware that you should re-download the data once 4.4.0 is released (Rscript start.R -> download data).
